### PR TITLE
RavenDB-20084: Fix backup status update issue for idle databases

### PR DESCRIPTION
**Pull request by ArieSLV**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20084/Backup-not-up-to-date-for-offline-and-not-modifed-db-that-was-not-loaded-yet-after-reboot

### Additional description

We encountered an issue with updating the status of the latest backup. If there were no changes in the database since the last backup, the next incremental backup would be skipped. Furthermore, if the database went into an idle state, there would be no update to the backup status.

This pull request addresses and resolves the aforementioned issue. We have introduced a new entity called IdleDatabaseActivity, which helps determine the activity that the timer, previously responsible for waking up the database, will receive. Now, upon timer expiration, either the database will be awakened or a simple update operation for the backup status will be performed.

Since a significant portion of the backup necessity determination mechanism was located in the PeriodicBackupRunner class instantiated in the DocumentDatabase class, which we unload, it was necessary to make the entire functionality static and move it to a static helper class called BackupUtils.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/Json/Serialization/JsonDeserializationClient.cs [hunk 1]
```diff
@@ -253,5 +253,7 @@ internal class JsonDeserializationClient : JsonDeserializationBase
         internal static readonly Func<BlittableJsonReaderObject, DatabaseSettings> DatabaseSettings = GenerateJsonDeserializationRoutine<DatabaseSettings>();
         
         public static readonly Func<BlittableJsonReaderObject, PutTrafficWatchConfigurationOperation.Parameters> GetTrafficWatchConfigurationResult = GenerateJsonDeserializationRoutine<PutTrafficWatchConfigurationOperation.Parameters>();
+        
+        public static readonly Func<BlittableJsonReaderObject, PeriodicBackupConfiguration> GetPeriodicBackupConfiguration = GenerateJsonDeserializationRoutine<PeriodicBackupConfiguration>();
```
Comment (grisha-kotler) [inline src/Raven.Client/Json/Serialization/JsonDeserializationClient.cs]: Move to `JsonDeserializationServer`

File: src/Raven.Server/Utils/BackupUtils.cs [hunk 2]
```diff
@@ -294,6 +305,182 @@ public static PathSetting GetBackupTempPath(RavenConfiguration configuration, st
...
+            // cluster is down
+            if (parameters.Logger.IsOperationsEnabled)
+                parameters.Logger.Operations($"Could not find the responsible node for backup task '{parameters.Configuration.TaskId}' of database '{parameters.DatabaseName}'.");
+
+            return new IdleDatabaseActivity(IdleDatabaseActivityType.WakeUpDatabase, DateTime.UtcNow);
+        }
+
+        if (responsibleNodeTag != parameters.ServerStore.NodeTag)
+        {
+            // not responsive for this backup task
```
Comment (grisha-kotler) [inline src/Raven.Server/Utils/BackupUtils.cs]:
```suggestion
            // not responsible for this backup task
```

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 3]
```diff
@@ -1050,18 +1063,23 @@ private void LogUnloadFailureReason(StringSegment databaseName, string reason)
                 _logger.Operations($"Could not unload database '{databaseName}', reason: {reason}");
         }
 
-        public void RescheduleDatabaseWakeup(string name, long milliseconds, DateTime? wakeup = null)
+        public void RescheduleNextIdleDatabaseActivity(string databaseName, IdleDatabaseActivity idleDatabaseActivity)
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs:1071]: Can we have only one overload of `RescheduleNextIdleDatabaseActivity`?

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 4]
```diff
@@ -1011,12 +1021,15 @@ public bool UnloadDirectly(StringSegment databaseName, DateTime? wakeup = null,
                 UnloadDatabaseInternal(databaseName.Value, caller);
                 LastRecentlyUsed.TryRemove(databaseName, out _);
 
-                if (dueTime > 0)
-                    _wakeupTimers.TryAdd(databaseName.Value, new Timer(_ => StartDatabaseOnTimer(databaseName.Value, wakeup), null, dueTime, Timeout.Infinite));
+                if (idleDatabaseActivity?.DueTime > 0)
+                    _wakeupTimers.TryAdd(databaseName.Value, new Timer(_ => NextScheduledActivityCallback(databaseName.Value, idleDatabaseActivity), null, idleDatabaseActivity.DueTime, Timeout.Infinite));
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: named parmater

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 5]
```diff
@@ -937,7 +939,9 @@ public DateTime LastWork(DocumentDatabase resource)
                     maxLastWork = env.Environment.LastWorkTime;
             }
 
-            return maxLastWork.AddMilliseconds(dbSize / 1024L);
+            return ForTestingPurposes is { ShouldFetchIdleStateImmediately: true }
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: Better add an `if` here (`ForTestingPurposes` is used only for tests)

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 6]
```diff
@@ -1070,20 +1088,48 @@ private void StartDatabaseOnTimer(string name, DateTime? wakeup = null)
...
+                            var backupStatus = BackupUtils.GetBackupStatusFromCluster(_serverStore, databaseName, nextIdleDatabaseActivity.TaskId);
+                            
+                            backupStatus.LastIncrementalBackup = backupStatus.LastIncrementalBackupInternal = nextIdleDatabaseActivity.TimeOfActivityUtc;
+                            backupStatus.LocalBackup.LastIncrementalBackup = nextIdleDatabaseActivity.TimeOfActivityUtc;
+                            backupStatus.LocalBackup.IncrementalBackupDurationInMs = 0;
+
+                            var backupResult = new BackupResult();
+                            backupResult.AddMessage($"Skipping incremental backup because no changes were made from last full backup on {backupStatus.LastFullBackup}.");
+
+                            BackupUtils.SaveBackupStatus(backupStatus, databaseName, _serverStore, _logger, backupResult);
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs:1107]: What happens if we weren't able to save the backup status?
When will it wake up again?

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 7]
```diff
@@ -1191,4 +1234,29 @@ private void CheckConfigurationPaths(RavenConfiguration parentConfiguration, Rav
             }
         }
     }
+
+    public class IdleDatabaseActivity
+    {
+        public long LastEtag { get; }
+        public IdleDatabaseActivityType IdleDatabaseActivityType { get; }
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]:
```suggestion
        public IdleDatabaseActivityType Type { get; }
```

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 8]
```diff
@@ -1191,4 +1234,29 @@ private void CheckConfigurationPaths(RavenConfiguration parentConfiguration, Rav
...
+        public DateTime? TimeOfActivityUtc { get; }
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]:
```suggestion
        public DateTime? DateTime { get; }
```

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 9]
```diff
@@ -1609,10 +1620,19 @@ private void LoadTimeSeriesPolicyRunnerConfigurations()
                         }
                     }
 
-                    if (ServerStore.LicenseManager.HasHighlyAvailableTasks() == false)
+                    if (serverStore.LicenseManager.HasHighlyAvailableTasks() == false)
                     {
                         // can't redistribute, keep it on the original node
-                        RaiseAlertIfNecessary(databaseTopology, configuration, lastResponsibleNode);
+                        if (databaseTopology.Count > 1 &&
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Why was it changed?

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 9 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: Returned to the previous implementation, adding a check for `notificationCenter` not being null.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 8 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: Done.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 7 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: Done.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 6 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs:1107]: We swallow all exceptions within the `BackupUtils.SaveBackupStatus` method.

Therefore, the time of the next awakening will be determined within the `BackupUtils.GetEarliestIdleDatabaseActivity` method and will coincide with the time of the next backup.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 5 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: Done.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 4 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: Done.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 3 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs:1071]: Done.

File: src/Raven.Server/Utils/BackupUtils.cs (hunk 2 above)
Comment (ArieSLV) [reply src/Raven.Server/Utils/BackupUtils.cs]: Done.

File: src/Raven.Client/Json/Serialization/JsonDeserializationClient.cs (hunk 1 above)
Comment (ArieSLV) [reply src/Raven.Client/Json/Serialization/JsonDeserializationClient.cs]: Done.

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 10]
```diff
@@ -1070,20 +1088,49 @@ private void StartDatabaseOnTimer(string name, DateTime? wakeup = null)
...
 
-                                  RescheduleDatabaseWakeup(name, milliseconds: _dueTimeOnRetry, wakeup);
-                              }
-                          }, TaskContinuationOptions.OnlyOnFaulted);
+                        case IdleDatabaseActivityType.WakeUpDatabase:
+                            _ = TryGetOrCreateResourceStore(databaseName, nextIdleDatabaseActivity.DateTime).ContinueWith(t =>
+                            {
+                                var ex = t.Exception.ExtractSingleInnerException();
+                                if (ex is DatabaseConcurrentLoadTimeoutException e)
+                                {
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs:1125]: Why do we need to handle this case?
The database will be loaded anyway, no?

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 11]
```diff
@@ -1576,9 +1576,20 @@ private void LoadTimeSeriesPolicyRunnerConfigurations()
             IDatabaseTask configuration,
             IDatabaseTaskStatus taskStatus,
             bool keepTaskOnOriginalMemberNode = false)
+        {
+            return WhoseTaskIsIt(ServerStore, databaseTopology, configuration, taskStatus, NotificationCenter, keepTaskOnOriginalMemberNode);
+        }
+
+        public static string WhoseTaskIsIt(
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Why do we need a new overload?
Can we move this to a helper class? And everyone will call it directly.

File: src/Raven.Server/Utils/BackupUtils.cs [hunk 12]
```diff
@@ -77,6 +81,13 @@ internal static BackupInfo GetBackupInfo(BackupInfoParameters parameters)
         };
     }
 
+    internal static PeriodicBackupStatus GetBackupStatusFromCluster(ServerStore serverStore, string databaseName, long taskId)
```
Comment (grisha-kotler) [inline src/Raven.Server/Utils/BackupUtils.cs]: private?

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Utils/BackupUtils.cs (hunk 12 above)
Comment (ArieSLV) [reply src/Raven.Server/Utils/BackupUtils.cs]: I have completely removed the overload of this method.

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 11 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: Done.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 10 above)
Comment (ArieSLV) [reply src/Raven.Server/Documents/DatabasesLandlord.cs:1125]: The decision to handle this specific exception type was made based on the following issue: https://issues.hibernatingrhinos.com/issue/RavenDB-18420/Database-might-not-wake-up-from-idle-if-we-are-waking-a-lot-of-databases-from-idle-state

**Review by grisha-kotler — APPROVED** [review]

**Pull request merged**